File: nc_150/linked_list/go/reverse_ll.go
// Iteration : just try to connect backwards as we move fwd

func reverseList(head *ListNode) *ListNode {
	curr := head
	var prev *ListNode

	for curr != nil {
		nxt := curr.Next
		curr.Next = prev
		prev = curr
		curr = nxt
	}
	return prev
}

// Recursive approach : this is a little tricky as it looks to connect while
// traversing from the back to the front

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	p := reverseList(head.Next)
	head.Next.Next = head
	head.Next = nil
	return p
}

File: nc_150/linked_list/reverse_ll.cpp
// Iteration : the idea is to simply keep connecting backwards as we go forward and return the 
// elem at the end 

// TC = O(n)
// SC = O(1)

ListNode* reverseList(ListNode* head) {
    ListNode* curr = head ;
    ListNode *nxt, *prev = nullptr;
    
    while(curr){
        nxt = curr -> next ;
        curr -> next = prev;
        prev = curr ;
        curr = nxt;
    }
    return prev;
}

// Recursive approach : try implementing it, 
// TC = O(n)
// SC = O(1)

ListNode* reverseList(ListNode* head) {
    if (head == NULL || head -> next == NULL)
        return head;

    ListNode* p = reverseList(head -> next);
    head ->next -> next = head;
    head -> next = NULL;
    return p;
}